daemon/internal/libcontainerd/remote/client.go

```diff
@@ -147,6 +147,39 @@ func (c *client) NewContainer(ctx context.Context, id string, ociSpec *specs.Spe
 	return &created, nil
 }
 
+// CleanupUnusedLeases removes leases that are not associated with any resources.
+// This will return the last error encountered while trying to delete leases.
+func (c *client) CleanupUnusedLeases(ctx context.Context) error {
+	leaseMan := c.client.LeasesService()
+	leases, err := leaseMan.List(ctx)
+	if err != nil {
+		return err
+	}
+
+	var lastErr error
+	for _, l := range leases {
+		// Skip leases that are still in use.
+		resources, err := leaseMan.ListResources(ctx, l)
+		if err != nil {
+			log.G(ctx).WithError(err).WithField("lease", l.ID).Debug("Failed to list containerd lease resources")
+			continue
+		}
+		if len(resources) > 0 {
```

Comment:
There is nothing in containerd that ensure that a leased resource is not deleted. This would account for cases where an non-expiring lease was created with a random ID, then never used against or anything ever added to it. Resources are never removed from a lease once they are added though.